pennylane/optimize/lie_algebra.py

```diff
@@ -0,0 +1,363 @@
+
+    Instantiate the optimizer with the initial circuit and the cost function and set the stepsize
+    accordingly:
+
+    >>> opt = qml.LieAlgebraOptimizer(circuit=quant_fun, stepsize=0.1)
+
+    Applying 5 steps gets us close the ground state of :math:`E\approx-2.23`:
+
+    >>> for step in range(6):
+    ...    circuit, cost = opt.step_and_cost()
+    ...    print(f"Step {step} - cost {cost}")
+    Step 0 - cost -1.3351865007304005
+    Step 1 - cost -1.9937887238935206
+    Step 2 - cost -2.1524234485729834
+    Step 3 - cost -2.1955105378898487
+    Step 4 - cost -2.2137628169764256
+    Step 5 - cost -2.2234364822091575
+
+    The optimized circuit is returned at each step, and can be used as any other QNode:
+
+    >>> circuit()
+    -2.2283086057521713
+
+    """
+    # pylint: disable=too-many-arguments
+    # pylint: disable=too-many-instance-attributes
+    def __init__(self, circuit, stepsize=0.01, restriction=None, exact=False, trottersteps=1):
+
+        if not isinstance(circuit, qml.QNode):
+            raise TypeError(f"circuit must be a QNode, received {type(circuit)}")
+
+        self.circuit = circuit
+        self.circuit.construct([], {})
+        self.hamiltonian = circuit.func().obs
+        if not isinstance(self.hamiltonian, qml.Hamiltonian):
+            raise TypeError(
+                f"circuit must return the expectation value of a Hamiltonian,"
+                f"received {type(circuit.func().obs)}"
+            )
+        self.nqubits = len(circuit.device.wires)
+
+        if self.nqubits > 4:
+            warnings.warn(
+                "The exact Riemannian gradient is exponentially expensive in the number of qubits,"
+                f"optimizing a {self.nqubits} qubit circuit may be slow.",
+                UserWarning,
+            )
+        if restriction is not None and not isinstance(restriction, qml.Hamiltonian):
+            raise TypeError(f"restriction must be a Hamiltonian, received {type(restriction)}")
+        (
+            self.lie_algebra_basis_ops,
+            self.lie_algebra_basis_names,
+        ) = self.get_su_n_operators(restriction)
+        self.exact = exact
+        self.trottersteps = trottersteps
+        self.coeffs, self.observables = self.hamiltonian.terms
+        self.stepsize = stepsize
+
+    def step(self):
+        r"""Update the circuit with one step of the optimizer.
+
+        Returns:
+           float: the optimized circuit and the objective function output prior
+           to the step.
+        """
+        return self.step_and_cost()[0]
+
+    def step_and_cost(self):
+        r"""Update the circuit with one step of the optimizer and return the corresponding
+        objective function value prior to the step.
+
+        Returns:
+            tuple[.QNode, float]: the optimized circuit and the objective function output prior
+            to the step.
+        """
+        # pylint: disable=not-callable
+
+        cost = self.circuit()
```

Comment:
@therooler this is my fault for not noticing this earlier, but:

I think it should be trivial to support parametrized QNodes by simply doing
```suggestion
    def step_and_cost(self, *args, **kwargs):
        r"""Update the circuit with one step of the optimizer and return the corresponding
        objective function value prior to the step.

        Returns:
            tuple[.QNode, float]: the optimized circuit and the objective function output prior
            to the step.
        """
        # pylint: disable=not-callable

        cost = self.circuit(*args, **kwargs)
```

(and similar for `self.step()`).

This should be a minor change, but potentially a big quality-of-life improvement for users, and bring the optimizer in line with other optimizers!

Reply:
(Note that this is not required for approval, as I have already approved)

Reply:
Let me see if this breaks things.

Reply:
So the issue here is that I need to construct the circuit in __init__, otherwise things break. I haven't been able to figure out how make things work without that.

Reply:
Oh of course, thanks for exploring this!